docs/src/threadpool.rst

```diff
@@ -32,35 +55,156 @@ Data types
 
 .. c:type:: void (*uv_work_cb)(uv_work_t* req)
 
-    Callback passed to :c:func:`uv_queue_work` which will be run on the thread
-    pool.
+    Callback passed to :c:func:`uv_queue_work` which will be run on
+    the executor.
 
 .. c:type:: void (*uv_after_work_cb)(uv_work_t* req, int status)
 
     Callback passed to :c:func:`uv_queue_work` which will be called on the loop
-    thread after the work on the threadpool has been completed. If the work
-    was cancelled using :c:func:`uv_cancel` `status` will be ``UV_ECANCELED``.
+    thread after the work on the executor has been completed or cancelled.
+    If the work was cancelled using :c:func:`uv_cancel`,
+    then `status` will be ``UV_ECANCELED``.
+
+.. c:type:: uv_executor_t
+
+    Executor type. Use when overriding the default threadpool.
+
+.. c:type:: void (*uv_executor_init_func)(uv_executor_t* executor)
+
+    Called once, the first time :c:func:`uv_executor_queue_work` is called.
+    Do any initialization you want.
+
+.. c:type:: void (*uv_executor_destroy_func)(uv_executor_t* executor)
+
+    Called at some point ???
+
+.. c:type:: void (*uv_executor_submit_func)(uv_executor_t* executor, uv_work_t* req, uv_work_options_t* opts)
 
+    Called when the executor should handle this request.
+
+.. c:type:: int (*uv_executor_cancel_func)(uv_executor_t* executor, uv_work_t* req)
+
+    Called when someone wants to cancel a previously-submit'ed request.
+    Return ``UV_EBUSY`` if you cannot cancel it.
+
+.. seealso:: :c:func:`uv_cancel_t`.
+
+.. c:type:: int (*uv_executor_done_cb)(uv_work_t* req)
+
+    libuv sets this during a successful call to :c:func:`uv_replace_executor`.
+    An executor should invoke this CB once finished with a request.
+    This CB is thread safe.
+
+.. c:type:: int uv_replace_executor(uv_executor_t* executor)
+
+    Replace the default libuv executor with this user-defined one.
+    Must be called before any work is submitted to the default libuv executor.
+    Returns 0 on successs.
```

Comment:
typo: `success`

Can we specify the lifetime requirements for `executor` here, and what happens if there is currently executing code? 

Also, is there any chance we could specify executors per-loop at some point? I’d guess that’s tricky because a lot of the code currently isn’t laid out for that?

Reply:
Perhaps I am a snakesss.

> what happens if there is currently executing code?

Once work has been submitted, subsequent calls to `uv_replace_executor` will return non-zero.

> Also, is there any chance we could specify executors per-loop at some point? I’d guess that’s tricky because a lot of the code currently isn’t laid out for that?

You would need to ensure the executor was drained or paused. If paused you would also need a way to transfer work from the old executor to the new one.

Out of curiosity, why would we want this? For online tuning purposes? (If so I think the burden would be on the embedder to figure that out.)

Reply:
I’m partially thinking about tuning purposes, yes, but also about Node’s own embedder scenarios; Node.js is not necessarily the only user of libuv APIs in a process, but we’re most likely going to provide our own executor.